esp/services/ws_workunits/ws_workunitsHelpers.cpp

```diff
@@ -3974,6 +4156,100 @@ void CWsWuFileHelper::readWUFile(const char* wuid, const char* workingFolder, Ws
+    StringBuffer unzipFolder;
+    if (unzipZAPFile(zapFileName, password, unzipFolder) != 0)
+        throw MakeStringException(ECLWATCH_INTERNAL_ERROR, "Failed to upzip ZAP File: %s.", zapFileName);
+
+    //Find ESP's IP and working folder
+    StringBuffer localIP, importFilePath, wuxmlFile;
+    IpAddress ipaddr = queryHostIP();
+    ipaddr.getIpText(localIP);
+    getConfigurationDirectory(directories, "run", "esp", espName, importFilePath);
+
+    //importFilePath: absolute path for WU files
+    //wuxmlFile: relative path for WU XML file
+    importFilePath.append(PATHSEPSTR).append(unzipFolder).append(PATHSEPSTR);
+    wuxmlFile.set(unzipFolder).append(PATHSEPSTR);
+    wuxmlFile.append(strlen(zapFileName) - 3, zapFileName).append("xml");
+
+    //Add WU files and update dali
+    if (!factory->importWorkUnit(wuid, zapFileName, wuxmlFile, localIP, importFilePath, importQueryAssociatedFile))
+        throw MakeStringException(ECLWATCH_INTERNAL_ERROR, "Failed to import workunit from ZAP File: %s. Check ESP log for details.", zapFileName);
+
+    //Import GraphProgress
+    importGraphProgress(wuid, zapFileName, unzipFolder);
+}
+
+int CWsWuFileHelper::unzipZAPFile(const char* zapFileName, const char* password, StringBuffer& outFolder)
+{
+    //Set/check/clean the outFolder for unzip
+    //Each ZAP WU has its subfolder under the workunitFolder. Use the zapFileName to create the subfolder.
+    outFolder.set(workunitFolder);
+    outFolder.append(strlen(zapFileName) - 4, zapFileName); //remove .zip from the zapFileName
+    Owned<IFile> unzipFolder = createIFile(outFolder);
+    if (!unzipFolder->exists())
+        unzipFolder->createDirectory();
+    else
+        cleanFolder(unzipFolder, false);
+
+    //Unzip ZAP file
+    StringBuffer zipCommand;
+    if (!isEmptyString(password))
+        zipCommand.setf("unzip %s%s -P %s -d %s", zipFolder, zapFileName, password, outFolder.str());
+    else
+        zipCommand.setf("unzip %s%s -d %s", zipFolder, zapFileName, outFolder.str());
+    return (system(zipCommand.str()));
+}
+
+void CWsWuFileHelper::importGraphProgress(const char* wuid, const char* zapFileName, const char* progressFileFolder)
+{
+    Owned<IRemoteConnection> conn = querySDS().connect("/GraphProgress", myProcessSession(), RTM_LOCK_WRITE, SDS_LOCK_TIMEOUT);
+    if (!conn)
+        throw MakeStringException(ECLWATCH_CANNOT_CONNECT_DALI, "Failed to connect to Dali for GraphProgress.");
+
+    Owned<IPropertyTree> graphProgressRoot = conn->getRoot();
+    if (graphProgressRoot->queryBranch(wuid))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to import GraphProgress from %s: workunit %s alreasy exists.", zapFileName, wuid);
+
+    //Read the GraphProgress from the .graphprogress file inside the ZAP report.
+    StringBuffer graphProgressFile;
+    graphProgressFile.set(progressFileFolder).append(PATHSEPSTR);
+    graphProgressFile.append(strlen(zapFileName) - 3, zapFileName).append("graphprogress");
```

Comment:
new_WUID.graphprogress?

Reply:
Once importWorkunit functionality is all within workunit.cpp - it will probably become clearer.
The progress side it will be less separate, and if using libarchive approach, you will not have filename at the time you're reading it.